### PR TITLE
docs: forward-ports from 3.x

### PR DESCRIPTION
**Pull request by tchaikov**
the document on 4.x branch does not build without this changeset. in order to build docs from 4.x branch, we have to have https://github.com/scylladb/java-driver/commit/cdaf75a6cf54b91f3f795aa509cf836166dad1e2. but since we are here, let's forward-port all 3.x doc change to 4.x. 

tested using 

```console
$ make setup
$ make preview
```

### CODE REVIEW

**Comment by roydahan** [comment]
Need to rebase PR in order to re-run the CI.

**Comment by tchaikov** [comment]
rebased and repushed.

**Comment by avelanarius** [comment]
@dgarcia360 @annastuchlik Please review this PR.

@tchaikov I'm not really sure I understand the reasoning for this PR. The cover letter doesn't really explain the intention.

**Comment by tchaikov** [comment]
> @dgarcia360 @annastuchlik Please review this PR.
> 
> @tchaikov I'm not really sure I understand the reasoning for this PR. The cover letter doesn't really explain the intention.

@avelanarius hi Piotr, the document on 4.x branch does not build without this changeset. that's why i ported the change to address the FTBFS.

**Comment by dgarcia360** [comment]
@annastuchlik @tchaikov @avelanarius  It looks good to me. The issue is that, without these changes, the 4.x will not build locally.

**Comment by roydahan** [comment]
Ping reviewers

File: .github/workflows/docs-pages.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: "Docs / Publish"
+# For more information,
+# see https://sphinx-theme.scylladb.com/stable/deployment/production.html#available-workflows
+
+on:
+  push:
+    branches:
+    - scylla-3.x
```
Comment (Bouncheck) [inline .github/workflows/docs-pages.yaml:8]: This feels wrong. Wouldn't that never fire?

File: .github/workflows/docs-pr.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+name: "Docs / Build PR"
+# For more information,
+# see https://sphinx-theme.scylladb.com/stable/deployment/production.html#available-workflows
+
+on:
+  pull_request:
+    branches:
+    - scylla-3.x
```
Comment (Bouncheck) [inline .github/workflows/docs-pr.yaml]: Similar as above. We should not have a workflow for PRs to 3.x on 4.x branch.

**Review by Bouncheck — CHANGES_REQUESTED** [review]
I successfully ran `make preview` after checking out this branch, however I've seen that script had to downgrade a lot of dependencies. Maybe a rebase is needed.
Github workflows seem incorrect or unnecessary: do we do docs deploy from 4.x branch? Deploy.sh seems not to be there anyway.
Is this purely a port? CustomCommonMarkParser seems to be new, do we not need it in 3.x?
Branchest/Tags need an update: latest tag is 4.18.0.1

**Comment by Bouncheck** [comment]
How do we maintain those two docs branches going forward? Is it possible to make it so we keep up to date setup on `scylla-3.x` and from time to time copy whole `docs` subdirectory to `scylla-4.x`?

**Comment by tchaikov** [comment]
> I successfully ran `make preview` after checking out this branch, however I've seen that script had to downgrade a lot of dependencies. Maybe a rebase is needed. 

sure. rebased.

> Github workflows seem incorrect or unnecessary: do we do docs deploy from 4.x branch? 

i didn't manually edit the backport commits unless i had to to resolve the conflicts or to fix the document build. the purpose of this changeset is to enable us to build document on 3.x branch. i am not sure if we deploy docs from 4.x branch. guess this is a question to the maintainer of this project.

considering this PR has been around for half a year, can we get this changeset reviewed and then work on the CI of document deployment if this is desirable?

> Deploy.sh seems not to be there anyway. Is this purely a port? 

yes, it is a backport. but again, this PR has been around for ages. and i haven't cherry-picked new changes from 4.x since i sent it for review. if something is odd. the reason is very likely that it's caused by merge conflict or missing commits from 4.x. 

> CustomCommonMarkParser seems to be new, do we not need it in 3.x? 

on the contrary, it is old. the change removing `CustomCommonMarkParser` landed in 9eedcaca09d1bb5bb87452389c1eb9cc1bd9645f 

> Branchest/Tags need an update: latest tag is 4.18.0.1

could you be more specific? which files shall i update?

**Comment by tchaikov** [comment]
v2:

- cherry-pick more changes to `docs/source/conf.py`  from 4.x 
- suppress warning from autosectionlabel
- correct the codeblock directive (the patch was also sent to upstream, see https://github.com/apache/cassandra-java-driver/pull/1940)

**Comment by tchaikov** [comment]
> How do we maintain those two docs branches going forward?

again, i think it's up to the maintainer of this project.

>  Is it possible to make it so we keep up to date setup on `scylla-3.x` and from time to time copy whole `docs` subdirectory to `scylla-4.x`?

i don't think to "copy" the whole `docs` subdirectory to `scylla-4.x` branch is the right approach. my 2 cents would be

1. get the 4.x doc build in shape, so that it can at least build
2. decide if / how to deploy it
3. setup the CI to backport and deploy it
4. add the proper label when sending doc patches to 3.x, and let CI to backport the change.

**Comment by tchaikov** [comment]
@Bouncheck hi Wojciech, could you please take another look?

**Comment by Bouncheck** [comment]
`make clean preview` looks better. Now manual pages can be browsed normally (before they did not work). API Documentation does not link to javadoc.

`make clean multiversionpreview` does not build correctly for me. For testing I've removed all but 2 versions (4.14.1.x and 4.15.0.x) from `BRANCHES` in source/conf.py to speed up the build.  I'm getting a 404 error page instead.
Throughout the build I'm getting a lot of lines like this one, after "updating environment" step:
```
updating environment: [new config] 102 added, 0 changed, 0 removed
/home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
  warn("Container node skipped: type={0}".format(mdnode.t))
/home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
  warn("Container node skipped: type={0}".format(mdnode.t))
```
Visiting the `127.0.0.1:5500` directs me to `http://127.0.0.1:5500/master` which has 404 error. However pages like `http://127.0.0.1:5500/stable/index.html` or `http://127.0.0.1:5500/scylla-4.15.0.x/` remain functional with redirects to javadoc when clicking on API Documentation header too.

> >    Branchest/Tags need an update: latest tag is 4.18.0.1
>
>could you be more specific? which files shall i update?

I was guessing that we'd need to just add '4.18.0.1' to `TAGS` in docs/source/conf.py but that actually breaks the build with multiple warnings like:
```
/tmp/tmpk4aq9ak4/970452ef5b72d76b4f783335fccabe57107d9735/docs/source/manual/core/pooling/index.md: WARNING: document isn't included in any toctree
```
So after all let's not do that for now.

**Comment by tchaikov** [comment]
> `make clean preview` looks better. Now manual pages can be browsed normally (before they did not work). API Documentation does not link to javadoc.
> 
> `make clean multiversionpreview` does not build correctly for me. For testing I've removed all but 2 versions (4.14.1.x and 4.15.0.x) from `BRANCHES` in source/conf.py to speed up the build. I'm getting a 404 error page instead. Throughout the build I'm getting a lot of lines like this one, after "updating environment" step:
> 

i commented out "multiversionpreview" to workaround a dependency issue on python 3.12, but forgot to revert it. the latest version dropped this workaround. so the multiversion should be back now.

> ```
> updating environment: [new config] 102 added, 0 changed, 0 removed
> /home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
>   warn("Container node skipped: type={0}".format(mdnode.t))
> /home/wo/.cache/pypoetry/virtualenvs/java-driver-xPfC9vS5-py3.11/lib/python3.11/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
>   warn("Container node skipped: type={0}".format(mdnode.t))
> ```

we have the same error on 3.x branch.

> 
> Visiting the `127.0.0.1:5500` directs me to `http://127.0.0.1:5500/master` which has 404 error. However pages like `http://127.0.0.1:5500/stable/index.html` or `http://127.0.0.1:5500/scylla-4.15.0.x/` remain functional with redirects to javadoc when clicking on API Documentation header too.
> 

i think this is due to the removed multiversion in previous reversion. and should be fixed.

> > > Branchest/Tags need an update: latest tag is 4.18.0.1
> > 
> > 
> > could you be more specific? which files shall i update?
> 
> I was guessing that we'd need to just add '4.18.0.1' to `TAGS` in docs/source/conf.py but that actually breaks the build with multiple warnings like:
> 

yeah, all 4.x tags are broken at this moment. i think the first step is to fix the 4.x branch.

> ```
> /tmp/tmpk4aq9ak4/970452ef5b72d76b4f783335fccabe57107d9735/docs/source/manual/core/pooling/index.md: WARNING: document isn't included in any toctree
> ```
> 
> So after all let's not do that for now.

**Comment by tchaikov** [comment]
v3:

- revert the change to remove multiversion support.

**Comment by tchaikov** [comment]
@Bouncheck i see you are still asking for changes. may i learn what exactly the changes i need to make?

**Comment by Bouncheck** [comment]
I'm not getting redirected to javadoc when clicking on API Documentation on generated page (`make clean preview`).
But if that's intended for now I can approve anyway since this PR is a big improvement.
I haven't checked this on 3.x branch but if it's the same there it would make sense to me to approve now and have a separate fix for 3.x + port to 4.x later on.

**Comment by tchaikov** [comment]
> I'm not getting redirected to javadoc when clicking on API Documentation on generated page (`make clean preview`). 

@Bouncheck to my best knowledge, this is indeed intended. see https://github.com/scylladb/java-driver/blob/9eedcaca09d1bb5bb87452389c1eb9cc1bd9645f/docs/source/conf.py#L174-L179

we redirect the api doc to `${version_name}/api/index.html`, and if `version_name` is 14.5.6, then the link points to `/14.5.6/api/index.html`, right under the root directory. i think this should work if the html is deployed to a web server.

> But if that's intended for now I can approve anyway since this PR is a big improvement. I haven't checked this on 3.x branch but if it's the same there it would make sense to me to approve now and have a separate fix for 3.x + port to 4.x later on.

i'd appreciate if you could help approve and merge this change. based on this change, we would be able to improve the 4.x document further.

**Review by Bouncheck — APPROVED** [review]

**Pull request merged**